File: csharp/Carballo/Com/Alonsoruibal/Chess/Movegen/MagicMoveGenerator.cs
/*
 * Code converted with Sharpen
 * 
 */using Com.Alonsoruibal.Chess;
using Com.Alonsoruibal.Chess.Bitboard;
using Com.Alonsoruibal.Chess.Movegen;
using Sharpen;

namespace Com.Alonsoruibal.Chess.Movegen
{
	/// <summary>
	/// Magic move generator
	/// Pseudo-legal Moves
	/// </summary>
	/// <author>Alberto Alonso Ruibal</author>
	public class MagicMoveGenerator : MoveGenerator
	{
		private int[] moves;

		private int moveIndex;

		private long all;

		private long mines;

		private long others;

		internal BitboardAttacks bbAttacks;

		public virtual int GenerateMoves(Board board, int[] moves, int mIndex)
		{
			this.moves = moves;
			bbAttacks = BitboardAttacks.GetInstance();
			moveIndex = mIndex;
			all = board.GetAll();
			// only for clearity
			mines = board.GetMines();
			others = board.GetOthers();
			byte index = 0;
			long square = unchecked((long)(0x1L));
			while (square != 0)
			{
				if (board.GetTurn() == ((square & board.whites) != 0))
				{
					if ((square & board.rooks) != 0)
					{
						// Rook
						GenerateMovesFromAttacks(Move.ROOK, index, bbAttacks.GetRookAttacks(index, all));
					}
					else
					{
						if ((square & board.bishops) != 0)
						{
							// Bishop
							GenerateMovesFromAttacks(Move.BISHOP, index, bbAttacks.GetBishopAttacks(index, all
								));
						}
						else
						{
							if ((square & board.queens) != 0)
							{
								// Queen
								GenerateMovesFromAttacks(Move.QUEEN, index, bbAttacks.GetRookAttacks(index, all));
								GenerateMovesFromAttacks(Move.QUEEN, index, bbAttacks.GetBishopAttacks(index, all
									));
							}
							else
							{
								if ((square & board.kings) != 0)
								{
									// King
									GenerateMovesFromAttacks(Move.KING, index, bbAttacks.king[index]);
								}
								else
								{
									if ((square & board.knights) != 0)
									{
										// Knight
										GenerateMovesFromAttacks(Move.KNIGHT, index, bbAttacks.knight[index]);
									}
									else
									{
										if ((square & board.pawns) != 0)
										{
											// Pawns
											if ((square & board.whites) != 0)
											{
												if (((square << 8) & all) == 0)
												{
													AddMoves(Move.PAWN, index, index + 8, (square << 8), false, true, 0);
													// Two squares if it is in he first row	
													if (((square & BitboardUtils.b2_d) != 0) && (((square << 16) & all) == 0))
													{
														AddMoves(Move.PAWN, index, index + 16, (square << 16), false, false, 0);
													}
												}
												GeneratePawnCapturesFromAttacks(index, bbAttacks.pawnUpwards[index], board.GetPassantSquare
													());
											}
											else
											{
												if ((((long)(((ulong)square) >> 8)) & all) == 0)
												{
													AddMoves(Move.PAWN, index, index - 8, ((long)(((ulong)square) >> 8)), false, true
														, 0);
													// Two squares if it is in he first row	
													if (((square & BitboardUtils.b2_u) != 0) && ((((long)(((ulong)square) >> 16)) & all
														) == 0))
													{
														AddMoves(Move.PAWN, index, index - 16, ((long)(((ulong)square) >> 16)), false, false
															, 0);
													}
												}
												GeneratePawnCapturesFromAttacks(index, bbAttacks.pawnDownwards[index], board.GetPassantSquare
													());
											}
										}
									}
								}
							}
						}
					}
				}
				square <<= 1;
				index++;
			}
			square = board.kings & mines;
			// my king
			byte myKingIndex = unchecked((byte)(-1));
			// Castling: disabled when in check or squares attacked
			if ((((all & (board.GetTurn() ? unchecked((long)(0x06L)) : unchecked((long)(0x0600000000000000L
				)))) == 0 && (board.GetTurn() ? board.GetWhiteKingsideCastling() : board.GetBlackKingsideCastling
				()))))
			{
				myKingIndex = BitboardUtils.Square2Index(square);
				if (!board.GetCheck() && !bbAttacks.IsIndexAttacked(board, unchecked((byte)(myKingIndex
					 - 1)), board.GetTurn()) && !bbAttacks.IsIndexAttacked(board, unchecked((byte)(myKingIndex
					 - 2)), board.GetTurn()))
				{
					AddMoves(Move.KING, myKingIndex, myKingIndex - 2, 0, false, false, Move.TYPE_KINGSIDE_CASTLING
						);
				}
			}
			if ((((all & (board.GetTurn() ? unchecked((long)(0x70L)) : unchecked((long)(0x7000000000000000L
				)))) == 0 && (board.GetTurn() ? board.GetWhiteQueensideCastling() : board.GetBlackQueensideCastling
				()))))
			{
				if (myKingIndex == -1)
				{
					myKingIndex = BitboardUtils.Square2Index(square);
				}
				if (!board.GetCheck() && !bbAttacks.IsIndexAttacked(board, unchecked((byte)(myKingIndex
					 + 1)), board.GetTurn()) && !bbAttacks.IsIndexAttacked(board, unchecked((byte)(myKingIndex
					 + 2)), board.GetTurn()))
				{
					AddMoves(Move.KING, myKingIndex, myKingIndex + 2, 0, false, false, Move.TYPE_QUEENSIDE_CASTLING
						);
				}
			}
			return moveIndex;
		}

		/// <summary>Generates moves from an attack mask</summary>
		private void GenerateMovesFromAttacks(int pieceMoved, int fromIndex, long attacks
			)
		{
			while (attacks != 0)
			{
				long to = BitboardUtils.Lsb(attacks);
				// If we collide with other piece (or other piece and cannot capture), this is blocking
				if ((to & mines) == 0)
				{
					// Capturing
					AddMoves(pieceMoved, fromIndex, BitboardUtils.Square2Index(to), to, ((to & others
						) != 0), true, 0);
				}
				attacks ^= to;
			}
		}

		private void GeneratePawnCapturesFromAttacks(int fromIndex, long attacks, long passant
			)
		{
			while (attacks != 0)
			{
				long to = BitboardUtils.Lsb(attacks);
				if ((to & others) != 0)
				{
					AddMoves(Move.PAWN, fromIndex, BitboardUtils.Square2Index(to), to, true, true, 0);
				}
				else
				{
					if ((to & passant) != 0)
					{
						AddMoves(Move.PAWN, fromIndex, BitboardUtils.Square2Index(to), to, true, true, Move
							.TYPE_PASSANT);
					}
				}
				attacks ^= to;
			}
		}

		/// <summary>
		/// Adds an operation
		/// to onlyneeded for captures
		/// </summary>
		private void AddMoves(int pieceMoved, int fromIndex, int toIndex, long to, bool capture
			, bool checkPromotion, int moveType)
		{
			if (checkPromotion && (pieceMoved == Move.PAWN) && ((to & (BitboardUtils.b_u | BitboardUtils
				.b_d)) != 0))
			{
				moves[moveIndex++] = Move.GenMove(fromIndex, toIndex, pieceMoved, capture, Move.TYPE_PROMOTION_QUEEN
					);
				moves[moveIndex++] = Move.GenMove(fromIndex, toIndex, pieceMoved, capture, Move.TYPE_PROMOTION_KNIGHT
					);
				moves[moveIndex++] = Move.GenMove(fromIndex, toIndex, pieceMoved, capture, Move.TYPE_PROMOTION_ROOK
					);
				moves[moveIndex++] = Move.GenMove(fromIndex, toIndex, pieceMoved, capture, Move.TYPE_PROMOTION_BISHOP
					);
			}
			else
			{
				moves[moveIndex++] = Move.GenMove(fromIndex, toIndex, pieceMoved, capture, moveType
					);
			}
		}
	}
}

File: core/src/main/java/com/alonsoruibal/chess/movegen/MagicMoveGenerator.java
package com.alonsoruibal.chess.movegen;

import com.alonsoruibal.chess.Board;
import com.alonsoruibal.chess.Color;
import com.alonsoruibal.chess.Move;
import com.alonsoruibal.chess.Piece;
import com.alonsoruibal.chess.bitboard.AttacksInfo;
import com.alonsoruibal.chess.bitboard.BitboardUtils;

/**
 * Magic move generator
 * Generate pseudo-legal moves because can leave the king in check.
 * It does not set the check flag.
 *
 * @author Alberto Alonso Ruibal
 */
public class MagicMoveGenerator implements MoveGenerator {

	private int[] moves;
	private int moveIndex;
	private long all;
	private long mines;
	private long others;

	private final AttacksInfo ai = new AttacksInfo();

	public int generateMoves(Board board, int[] moves, int startIndex) {
		this.moves = moves;
		ai.build(board);

		moveIndex = startIndex;
		all = board.getAll();
		mines = board.getMines();
		others = board.getOthers();

		int index = 0;
		long square = 0x1L;
		while (square != 0) {
			if (board.getTurn() == ((square & board.whites) != 0)) {
				if ((square & board.rooks) != 0) { // Rook
					generateMovesFromAttacks(Piece.ROOK, index, ai.attacksFromSquare[index] & ~mines);
				} else if ((square & board.bishops) != 0) { // Bishop
					generateMovesFromAttacks(Piece.BISHOP, index, ai.attacksFromSquare[index] & ~mines);
				} else if ((square & board.queens) != 0) { // Queen
					generateMovesFromAttacks(Piece.QUEEN, index, ai.attacksFromSquare[index] & ~mines);
				} else if ((square & board.kings) != 0) { // King
					generateMovesFromAttacks(Piece.KING, index, ai.attacksFromSquare[index] & ~mines);
				} else if ((square & board.knights) != 0) { // Knight
					generateMovesFromAttacks(Piece.KNIGHT, index, ai.attacksFromSquare[index] & ~mines);
				} else if ((square & board.pawns) != 0) { // Pawns
					if ((square & board.whites) != 0) {
						if (((square << 8) & all) == 0) {
							addMoves(Piece.PAWN, index, index + 8, false, 0);
							// Two squares if it is in he first row	
							if (((square & BitboardUtils.b2_d) != 0) && (((square << 16) & all) == 0)) {
								addMoves(Piece.PAWN, index, index + 16, false, 0);
							}
						}
						generatePawnCapturesFromAttacks(index, ai.attacksFromSquare[index], board.getPassantSquare());
					} else {
						if (((square >>> 8) & all) == 0) {
							addMoves(Piece.PAWN, index, index - 8, false, 0);
							// Two squares if it is in he first row	
							if (((square & BitboardUtils.b2_u) != 0) && (((square >>> 16) & all) == 0)) {
								addMoves(Piece.PAWN, index, index - 16, false, 0);
							}
						}
						generatePawnCapturesFromAttacks(index, ai.attacksFromSquare[index], board.getPassantSquare());
					}
				}
			}
			square <<= 1;
			index++;
		}

		// Castling: disabled when in check or king route attacked
		if (!board.getCheck()) {
			int us = board.getTurn() ? Color.W : Color.B;

			long kingCastlingDestination = board.canCastleKingSide(us, ai);
			if (kingCastlingDestination != 0) {
				addMoves(Piece.KING, ai.kingIndex[us], BitboardUtils.square2Index(kingCastlingDestination), false, Move.TYPE_KINGSIDE_CASTLING);
			}
			long queenCastlingDestination = board.canCastleQueenSide(us, ai);
			if (queenCastlingDestination != 0) {
				addMoves(Piece.KING, ai.kingIndex[us], BitboardUtils.square2Index(queenCastlingDestination), false, Move.TYPE_QUEENSIDE_CASTLING);
			}
		}

		return moveIndex;
	}

	/**
	 * Generates moves from an attack mask
	 */
	private void generateMovesFromAttacks(int pieceMoved, int fromIndex, long attacks) {
		while (attacks != 0) {
			long to = Long.lowestOneBit(attacks);
			addMoves(pieceMoved, fromIndex, BitboardUtils.square2Index(to), ((to & others) != 0), 0);
			attacks ^= to;
		}
	}

	private void generatePawnCapturesFromAttacks(int fromIndex, long attacks, long passant) {
		while (attacks != 0) {
			long to = Long.lowestOneBit(attacks);
			if ((to & others) != 0) {
				addMoves(Piece.PAWN, fromIndex, BitboardUtils.square2Index(to), true, 0);
			} else if ((to & passant) != 0) {
				addMoves(Piece.PAWN, fromIndex, BitboardUtils.square2Index(to), true, Move.TYPE_PASSANT);
			}
			attacks ^= to;
		}
	}

	/**
	 * Adds a move (it can add a non legal move)
	 */
	private void addMoves(int pieceMoved, int fromIndex, int toIndex, boolean capture, int moveType) {
		if (pieceMoved == Piece.PAWN && (toIndex < 8 || toIndex >= 56)) {
			moves[moveIndex++] = Move.genMove(fromIndex, toIndex, pieceMoved, capture, Move.TYPE_PROMOTION_QUEEN);
			moves[moveIndex++] = Move.genMove(fromIndex, toIndex, pieceMoved, capture, Move.TYPE_PROMOTION_KNIGHT);
			moves[moveIndex++] = Move.genMove(fromIndex, toIndex, pieceMoved, capture, Move.TYPE_PROMOTION_ROOK);
			moves[moveIndex++] = Move.genMove(fromIndex, toIndex, pieceMoved, capture, Move.TYPE_PROMOTION_BISHOP);
		} else {
			moves[moveIndex++] = Move.genMove(fromIndex, toIndex, pieceMoved, capture, moveType);
		}
	}
}